### ISSUE TITLE
Graph Neural Networks for Web Page Understanding

### ISSUE THREAD

**Issue by johnhnguyen97** [Nguyenetic/argus#5]
## Overview

Implement Graph Neural Networks (GNN) to understand web page structure and extract content intelligently.

## Goals
- Convert DOM trees to graph representations
- Classify page elements (content, navigation, ads, boilerplate)
- Achieve >90% extraction accuracy
- Support zero-shot classification of new page types

### DOM to Graph Conversion
- **Nodes**: HTML elements (div, p, a, etc.)
- **Node Features**: 
  - Tag type (one-hot encoded)
  - CSS classes (embedded)
  - Text length
  - Position (x, y, width, height)
  - Styling attributes
- **Edges**:
  - Parent-child relationships
  - Sibling relationships
  - Hyperlink relationships

### GNN Model
- **Architecture**: GraphSAGE or GCN
- **Layers**: 3-4 graph convolutional layers
- **Output**: Node classification (content, navigation, ads, boilerplate)
- **Framework**: PyTorch Geometric or rust-ml

### Data Collection
- [ ] Scrape diverse websites for training data
- [ ] Manually label page elements (content vs navigation)
- [ ] Create benchmark dataset (1000+ pages)
- [ ] Split train/val/test sets (70/15/15)

### Graph Construction
- [ ] Parse HTML to DOM tree
- [ ] Extract node features from elements
- [ ] Build adjacency matrix
- [ ] Normalize features
- [ ] Convert to PyG Data format

### Model Training
- [ ] Implement GraphSAGE model
- [ ] Add training loop with Adam optimizer
- [ ] Implement early stopping
- [ ] Track accuracy, precision, recall, F1
- [ ] Save best model checkpoints

### Inference Pipeline
- [ ] Load trained model
- [ ] Convert new pages to graphs
- [ ] Run inference
- [ ] Extract classified content
- [ ] Filter out boilerplate/ads

### Integration
- [ ] Add `--gnn` flag to CLI
- [ ] Implement Python-Rust bridge (PyO3)
- [ ] Add model download/caching
- [ ] Benchmark extraction accuracy

## Example Usage
```bash
# Use GNN for intelligent extraction
argus scrape https://example.com --gnn

# Train custom GNN model
argus train-gnn --dataset ./labeled_pages --epochs 100
```

## Performance Targets
- Extraction accuracy: >90%
- Inference time: <50ms per page
- Model size: <100MB
- Precision/Recall: >0.85

## Dataset Examples
- News articles (NYT, BBC, CNN)
- Documentation sites (MDN, Rust docs)
- E-commerce (Amazon, eBay)
- Blogs (Medium, WordPress)

## References
- Week 7-8 in DETAILED_ROADMAP_RUST.md
- GraphSAGE paper: https://arxiv.org/abs/1706.02216
- PyTorch Geometric: https://pytorch-geometric.readthedocs.io/

**Phase:** 2 - Intelligence (Week 7-8)
**Priority:** Medium
